spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -444,6 +444,52 @@ protected void handleMessageInternal(Message<?> message) throws Exception {
 		}
 	}
 
+	protected boolean isExpireGroupsUponCompletion() {
+		return false;
+	}
+
+	private void removeEmptyGroupAfterTimeout(MessageGroup messageGroup, long timeout) {
+		Object groupId = messageGroup.getGroupId();
+		UUID groupUuid = UUIDConverter.getUUID(groupId);
+		ScheduledFuture<?> scheduledFuture = getTaskScheduler()
+				.schedule(() -> {
+					Lock lock = this.lockRegistry.obtain(groupUuid.toString());
+
+					try {
+						lock.lockInterruptibly();
+						try {
+							this.expireGroupScheduledFutures.remove(groupUuid);
+							MessageGroup groupNow = this.messageStore.getMessageGroup(groupUuid);
+							boolean removeGroup = groupNow.getLastModified()
+									<= (System.currentTimeMillis() - this.minimumTimeoutForEmptyGroups);
+							if (removeGroup) {
```

Comment:
Since we're checking this, I wonder if we should also check if it's still empty.

Also, what happens if removeGroup is false here - when will the group ever be removed?